### ISSUE TITLE
ipywidget based packages don't all work properly via thebelab on LibreTexts

### ISSUE THREAD

**Issue by moorepants** [LibreTexts/metalc#136]
We've see issues with:

- ipywidgets (e.g. can't make a slider control a plot)
- can't use the buttons on the matplotlib widget
- list other issues here

These are the widgets we'd like to support (check the box once installed and tested that it fully works):

- [x] https://github.com/matplotlib/ipympl
- [x] ipyleaflet
- [x] pythreejs
- [x] https://github.com/jupyter-widgets/ipywidgets
- [x] https://github.com/bloomberg/bqplot
- [x] https://github.com/maartenbreddels/ipyvolume
- [x] https://github.com/arose/nglview
- [x] https://github.com/QuantStack/ipycytoscape

**Comment by moorepants** [comment]
Magali Billen is working on a LibreTexts for GEL 56 and wants to use ipywidgets for some pages. Her student is going to report what is not working. I'm going to bump this to high priority so that we put some cycles into fixing this.

**Comment by nicolas-desalvio** [comment]
I'm Magali's student referenced in the previous comment. I'm trying to use Intslider from ipywidgets to change parameters and automatically update a figure. The sliders do not show up on LibreTexts, and there is no error message. 

The code and Libretexts page is here: https://geo.libretexts.org/Courses/University_of_California_Davis/UCD_GEL_56_-_Introduction_to_Geophysics/Geophysics_is_everywhere_in_geology.../1%3A_Rheology_of_Rocks/1.5_Inclined_Plane_Flow_Jupyter_Notebook

The picture below is what the same code does using Jupyter on my computer. When you change the sliders the graphs automatically update. So this is the goal.
![Slider_page1 5](https://user-images.githubusercontent.com/67663232/86183160-428cf580-bae6-11ea-91a3-b03c9a897322.png)

**Comment by moorepants** [comment]
Thanks for the example @nicolas-desalvio. We'll try to help you get it working.

**Comment by moorepants** [comment]
I pasted Nicolas's code into https://minrk.github.io/thebelab/ and got:

![image](https://user-images.githubusercontent.com/276007/86192222-e930c080-bafd-11ea-9768-09cafb6ba7ea.png)

So it may be that ThebeLab doesn't work with ipywidgets.

**Comment by moorepants** [comment]
Here is an open issue on ThebeLab:https://github.com/minrk/thebelab/issues/111

**Comment by sandertyu** [comment]
What's the backstory here @moorepants ? I've gathered a cursory understanding of what ThebeLab is, but I don't understand how it is interacting with libretext or where any of the code for the libretext integration is. 

According to that issue, a comment [here](https://github.com/minrk/thebelab/issues/111#issuecomment-457639508) claims to have fixed and ported the fix into ThebeLab. The issue was re-opened for something that doesn't seem relevant to this specific case here.

**Comment by moorepants** [comment]
The https://github.com/LibreTexts/ckeditor-binder-plugin uses Thebelab. Thebelab allows you to turn any given web page into a "binder powered" web page. It allows embedding of executable jupyter cells into the page. In our plugin, we load Thebelab and set configuration values to point it to my.binder.org and to various Dockerfiles (and associated built docker containers). All of the code is in the ckeditor plugin we wrote.

The ideal solution is to get a PR merged into Thebelab that fixes this issue for everyone using Thebelab. An alternative is for us to fork Thebelab and apply a patch, then use that (temporarily) until we can get a PR merged into Thebelab.

**Comment by moorepants** [comment]
One other historical note. We chose to use Thebelab instead of juniper (https://github.com/ines/juniper). I'd be curious to know if Juniper works with widgets. If so, maybe we should switch to Juniper...

**Comment by sandertyu** [comment]
Update on where we are currently at with ipywidgets:

We've found that on both libretexts and in the ckeditor plugin alone, trying to display ipywidgets sliders with code such as `ipywidgets.IntSlider()` will simply give an `Error displaying widget` message. However, on some ThebeLab pages such as [this](https://minrk.github.io/thebelab/) one, the widget will *always* display properly. This hints that at least part of the issue is in our implementation of ThebeLab in the ckeditor plugin, and I've made an [issue](https://github.com/minrk/thebelab/issues/226) over on the ThebeLab repository so that we can probably implement a similar fix. ***In the process of writing this, I went and tested ipywidgets on ThebeLab again. It worked on the [example](https://thebelab.readthedocs.io/en/latest/#examples) webpages where they previously had not, but then I ran it again 10 minutes later and it failed. I cannot say whether this will work when anyone else tries it. The widgets still always give `Error displaying widget` within our ckeditor plugin.***

The exact errors that we are having are identical to those explained and supposedly fixed in another issue on ThebeLab [here](https://github.com/minrk/thebelab/issues/72). At the bottom, one of the users mentions an output widget implementation, but there's no follow up. Similar conversations have gone on in other repositories like nbclient [here](https://github.com/jupyter/nbclient/pull/68) and [here](https://github.com/jupyter/nbclient/issues/39). You can really go down the rabbit hole with these issues if you follow the links to other pull requests/issues that they give and mention.

Furthermore, ThebeLab has had persistent issues with displaying widgets using interactive output. So there currently is no working way to use the `interact` method of displaying widgets within ThebeLab. In the mean time, I can search the ipywidget documentation to see if there is some other way that the student here can implement their code. Either way, we will have to update our ckeditor plugin so that it can display ipywidgets.

**Comment by mibillen** [comment]
Thank you so much for going down the rabbit hole to figure out what the underlying issue is. We have come up with a temporary work around for this using for loops, but it is much less useful for letting the student set values observe how the plots change.  I discussed this further with @nicolas-desalvio and he has another idea for a temporary work around if you can get the sliders working but not the interact option.

**Comment by nicolas-desalvio** [comment]
Here is an ipywidgets alternative if the widgets can be displayed but interact doesn't work

The first cell would have something like:

`import ipywidgets as widgets`
`h = widgets.IntSlider()`
`display(h)`

The next cell would just call the value of the slider similar to this:
`print(h.value)`

Instead of automatically updating, you have to rerun the second cell every time the slider changes.
Here's what it produces in ThebeLab:

![altslider](https://user-images.githubusercontent.com/67663232/87723914-5f285f00-c76f-11ea-8a0d-9dddf97aab46.png)

So this is a potential option Magali and I could use if the widgets can be displayed, but interact doesn't work.

**Comment by sandertyu** [comment]
That is one way you could do it, but unfortunately I am still waiting on a response from ThebeLab for how they have managed to get sliders to display on some of their webpages, so we cannot currently implement that workaround. In the mean time, another solution is instead to use [`ipywidgets.BoundedIntText()`](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html#BoundedIntText), which should take mostly the same arguments as `IntSlider()`, but the `layout` and `style` attributes may work differently. 

I've confirmed that this will display on libretexts, albeit you still cannot use it with `interact()`. Testing it myself, I was only able to get one `BoundedIntText()` to display per code cell, which I think is also an issue I have seen on ThebeLab. You may be able give the input boxes a cell each of their own, and then access their `value` and display the plot in another cell like you did above @nicolas-desalvio .

**Comment by moorepants** [comment]
From today's meeting: Noah and Kevin will try the different widget packages listed at the top of this issue in three environments: bare html thebelab, libretexts, and our jupyterhub cataloging (wit screenshots) what works and what doesn't. For the ones that work on libretexts create a page in the construction guide that shows off an example of them working.

**Comment by sandertyu** [comment]
We've been working on this, and we have put the code for our thebelab pages and a table for what does/does not work in each environment [here](https://github.com/sandertyu/widgetstuffs). You can see that below.

![image](https://user-images.githubusercontent.com/59950537/92191662-32143680-ee19-11ea-883a-04560e57c96e.png)

Would you like us to gather screenshots for each widget in each environment, or just screenshots from libretexts for the purpose of adding to the construction guide? We could do JupyterLab screenshots as well but most of these widgets already include some on their github and I'm not sure on which page we would put them anyways, unless  they are just for you to see.

**Comment by moorepants** [comment]
Screenshots for those that don't work is sufficient.

**Comment by nicolas-desalvio** [comment]
Thank you for all the work you've put into this.

Magali would like a general update as to where things stand with Ipywidgets (specifically interact and IntSlider). Has there been any progress since Noah's July 13th update on them?

**Comment by sandertyu** [comment]
Unfortunately, the status of ipywidgets has not changed since then. We rely on a backend Thebe to turn standard HTML text into executable code cells, the issue is multi-layered:

1. Our Binder plugin (the thing that lets you insert code cells on LibreTexts) does not properly display IntSlider() and some other widgets. There seems to be some way to get this to work on its own, but we need some guidance from the developers of Thebe themselves because it works very inconsistently. Even on their own pages it is usually broken.
2.  `interact()` simply does not work in Thebe at all. This will require some code changes on their end, which we hope to collaborate on. It's just slow going here unfortunately. Hopefully we are able to make some progress at the upcoming JupyterCon in October.
3. Our Binder plugin prevents cells from communicating properly. By this I mean that if you have code output in some cell, and you display a widget in a later cell, there is no way to interactively change the widget and automatically adjust the output of the first cell. This issue is 100% on our end, but even fixing this alone wouldn't let ipywidgets work properly.

We are sifting through layers of bugs with our Binder plugin, and hopefully we can reach the bottom of the stack where the above issues are fixed sometime during fall quarter, but we really do not know @nicolas-desalvio

**Comment by moorepants** [comment]
@nicolas-desalvio We are trying to organize a sprint for JupyterCon on Oct 17 to work on these issues. See: https://github.com/executablebooks/thebe/issues/239. Maybe you'd like to join us?

**Comment by nicolas-desalvio** [comment]
I'm sorry but I won't be able to join you @moorepants

**Comment by moorepants** [comment]
It could be worth seeing if these things work with juniper. If they do, then figuring out what juniper does that thebe doesn't could be helpful. (or maybe even switching to juniper if it all works correctly).

**Comment by celine168** [comment]
Juniper appears to be a more stripped down version of Thebe.
It doesn't import from @jupyter-widgets (comparing [Juniper's source code](https://github.com/ines/juniper/blob/master/src/index.js) with [Thebe's](https://github.com/executablebooks/thebe/blob/master/src/thebelab.js)) like Thebe does, which I suspect is why many Jupyter widgets do not appear in Juniper.

For example:
ipycytoscape:
![image](https://user-images.githubusercontent.com/22624020/99428418-78771e80-28bb-11eb-9c6f-a03d3fad9415.png)
bqplot:
![image](https://user-images.githubusercontent.com/22624020/99428479-8d53b200-28bb-11eb-93bb-7d4662932dcb.png)

**Comment by mandeepika** [comment]
> We've been working on this, and we have put the code for our thebelab pages and a table for what does/does not work in each environment in the checklist section of the README.md [here](https://github.com/sandertyu/widgetstuffs).

So, is this an updated table and all of the things listed on this table still don't work? Was taking a look at this issue, but not sure if these are still the widgets that don't work (since this comment is from Sept 2020).

**Comment by moorepants** [comment]
@sandertyu created the widgetstuffs repo, the state of functionality likely hasn't changed much. But we have addressed a few bugs since that, so maybe some work now. You'll have to confer with him and/or check to see what does and doesn't work.

After he created that we started adding examples to thebe, see https://thebe.readthedocs.io/en/latest/#examples. This is the first step to fixing things, i.e. ensure it works in thebe (alone). Here are some open PRs on thebe that are outstanding:

- https://github.com/executablebooks/thebe/issues/284
- https://github.com/executablebooks/thebe/pull/330
- https://github.com/executablebooks/thebe/pull/262
- https://github.com/executablebooks/thebe/issues/253
- https://github.com/executablebooks/thebe/issues/226
- and probably more

Once you are confident a package works with the latest thebe, then you can check to see if it runs on libretexts. If it doesn't you'll need to investigate if there are js/css that isn't loaded or if the loaded js/css is conflicting with something already on libretexts.

We should ultimately have pages in the libretexts constructions guide that demonstrate these different libraries working on libretexts, just like the examples in the thebe docs.

**Comment by moorepants** [comment]
I just dropped code examples for most of the packages that support visualization in jupyter here:

https://query.libretexts.org/Sandboxes/jupyterteam@ucdavis.edu

It give a single page to run to see if things are working or not. Seems like ipycytoscpae and pythreejs are the non-functional ones at the moment.

**Comment by moorepants** [comment]
Actually, we should probably do the same thing in the construction guide (but only display ones that work). Having a single page is much nicer and easier to check.

**Comment by sandertyu** [comment]
Big progress on the ipywidgets in Thebe, it looks like the current master branch allows for the original code posted by Nicolas to fully display;

![Thebe](https://cdn.discordapp.com/attachments/729801592433410138/829927851511775242/unknown.png)

However, moving the sliders does not activate the plot, so there's still more to be done. I think Celine's [PR](https://github.com/executablebooks/thebe/commit/beb37928a6d77cfb129b0192f9aa952130f64a2e) helped solve some things.

Furthermore, the Output widget example referenced [here](https://github.com/executablebooks/thebe/issues/226#issuecomment-814502484) also works properly now.

**Comment by moorepants** [comment]
Ok. I still have on my TODO to make the thebe release. I'll make it today or tomorrow.

**Comment by moorepants** [comment]
It seems the pythreejs animation is now working here: https://query.libretexts.org/Sandboxes/jupyterteam@ucdavis.edu, but I don't know what has changed. I thought it wouldn't even render. Now it does and the animation buttons work.

**Comment by mandeepika** [comment]
> It seems the pythreejs animation is now working here: https://query.libretexts.org/Sandboxes/jupyterteam@ucdavis.edu, but I don't know what has changed. I thought it wouldn't even render. Now it does and the animation buttons work.

Oh yeah you are right! I updated the page just now by uncommenting the DirectionalLight() function, which was originally the [issue](https://github.com/LibreTexts/widget-testing/issues/5). But it works now.

**Comment by moorepants** [comment]
That doesn't make any sense. Unless something on libretexts changed that removed some kind of clash with thebe. I thought pythreejs didn't work with thebe even on a bar html page.

**Comment by sandertyu** [comment]
I don't know the exact cause, but since y'all are looking at query, it seems likely that either Thebe 0.7.1, or the updated jupyterlab 3.0 and newer pythreejs in the default-env image used on query is allowing it to work. I've noticed that certain other widgets like ipyleaflet give some errors now though, but it's due to traitlet deprecation within their own codebase I think.

**Comment by TimStewartJ** [comment]
With the newest thebe release, it seems all of the original package we've been looking to support now work. An example of each can be found on https://query.libretexts.org/Sandboxes/jupyterteam_at_ucdavis.edu. Interestingly, the bokeh example doesn't work now? We'll have to do more testing with that.

**Comment by moorepants** [comment]
Bokeh has never worked with thebe.

**Comment by TimStewartJ** [comment]
That's what I suspected, I tried it with other versions and they didn't work either.

Is this issue good to close then?

**Comment by moorepants** [comment]
Are things live on production libretexts?

**Comment by TimStewartJ** [comment]
Ah not yet, I suppose it'd be best to wait until then.

**Comment by TimStewartJ** [comment]
TODO: Make/update construction guide pages (particularly https://chem.libretexts.org/Courses/Remixer_University/LibreTexts_Construction_Guide/05%3A_Interactive_Elements/5.02%3A_Jupyter_Notebooks_(Executable_Programming_Code_and_Figures)/5.2.08%3A_Using_ipywidgets_for_Interactivity)
